pkg/api/handlers/libpod/images.go

```diff
@@ -498,12 +503,95 @@ func CommitContainer(w http.ResponseWriter, r *http.Request) {
 	if len(query.Repo) > 0 {
 		destImage = fmt.Sprintf("%s:%s", query.Repo, tag)
 	}
-	commitImage, err := ctr.Commit(r.Context(), destImage, options)
+
+	if !query.Stream {
+		commitImage, err := ctr.Commit(r.Context(), destImage, options)
+		if err != nil && !strings.Contains(err.Error(), "is not running") {
+			utils.Error(w, http.StatusInternalServerError, fmt.Errorf("CommitFailure: %w", err))
+			return
+		}
+		utils.WriteResponse(w, http.StatusOK, entities.IDResponse{ID: commitImage.ID()})
+		return
+	}
+
+	// Channels all mux'ed in select{} below to follow API commit protocol
+	stdout := channel.NewWriter(make(chan []byte))
+	defer stdout.Close()
+	stderr := channel.NewWriter(make(chan []byte))
+	defer stderr.Close()
+	// Channels all mux'ed in select{} below to follow API commit protocol
+	options.CommitOptions.ReportWriter = stdout
+	var commitImage *libimage.Image
+	success := false
+	runCtx, cancel := context.WithCancel(r.Context())
+	go func() {
+		defer cancel()
+		commitImage, err = ctr.Commit(r.Context(), destImage, options)
+		if err == nil {
+			success = true
+		} else {
+			stderr.Write([]byte(err.Error() + "\n"))
+		}
+	}()
+
 	if err != nil && !strings.Contains(err.Error(), "is not running") {
 		utils.Error(w, http.StatusInternalServerError, fmt.Errorf("CommitFailure: %w", err))
 		return
 	}
-	utils.WriteResponse(w, http.StatusOK, entities.IDResponse{ID: commitImage.ID()})
+
+	flush := func() {
+		if flusher, ok := w.(http.Flusher); ok {
+			flusher.Flush()
+		}
+	}
+	w.WriteHeader(http.StatusOK)
+	w.Header().Set("Content-Type", "application/json")
+	flush()
```

Comment:
This happens too early.  At that point, we may already have an error from `Commit()`, so we should only send OK and the content-type on **first** successful stream.

Reply:
Moved to right after commit message above.